File: 13/nim/sol2.nim
import sets
import system/io
import strscans
import strutils

type Hole = tuple[y: int, x: int]

proc reflect(a: int, refl: bool, b: int): int =
  if refl and a > b: 2 * b - a else: a

# Read holes, reflect on folds
var holes: HashSet[Hole]
for line in stdin.lines:
  var x, y, pos: int
  var dim: char

  if scanf(line, "$i,$i", x, y):
    holes.incl((y: y, x: x))

  elif scanf(line, "fold along $c=$i", dim, pos):
    var new_holes: HashSet[Hole]
    for hole in holes.items:
      var hole = hole
      hole.x = hole.x.reflect(dim == 'x', pos)
      hole.y = hole.y.reflect(dim == 'y', pos)
      new_holes.incl(hole)
    holes = new_holes

# Print image, constructed by indices into string
var max_x, max_y = 0
for hole in holes:
  max_x = max_x.max(hole.x)
  max_y = max_y.max(hole.y)

var field = (' '.repeat(max_x + 1) & "\n").repeat(max_y + 1)

for hole in holes:
  var idx = hole.x + hole.y * (max_x + 2)
  field[idx] = '#'

echo field

File: 13/c/sol2.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

struct Hole {
  int x;
  int y;
} typedef Hole;

struct LL {
  void* data;
  struct LL* next;
} typedef LL;

void push (LL** next, void* data) {
  LL* head = malloc(sizeof(LL));
  head->data = data;
  head->next = *next;
  *next = head;
}

int main () {
  LL* holes = NULL;

  Hole* hole = malloc(sizeof(Hole));
  while (scanf("%d,%d\n", &hole->x, &hole->y) == 2) {
    push(&holes, hole);
    hole = malloc(sizeof(Hole));
  }

  char dim;
  int pos;
  while (scanf("fold along %c=%d\n", &dim, &pos) == 2) {
    LL* head = holes;
    while (head != NULL) {
      hole = (Hole*) head->data;
      switch (dim) {
        case 'x':
          if (hole->x > pos) hole->x = 2 * pos - hole->x;
          break;
        case 'y':
          if (hole->y > pos) hole->y = 2 * pos - hole->y;
          break;
      }
      head = head->next;
    }
  }

  int max_width = 0;
  int max_height = 0;

  LL* head = holes;
  while (head != NULL) {
    hole = (Hole*) head->data;
    if (max_width < hole->x) max_width = hole->x;
    if (max_height < hole->y) max_height = hole->y;
    head = head->next;
  }

  max_width++;
  max_height++;

  char* out = calloc(max_height * (max_width + 1) + 1, sizeof(char));
  memset(out, ' ', max_height * (max_width + 1));

  head = holes;
  while (head != NULL) {
    hole = (Hole*) head->data;
    out[hole->x + hole->y * (max_width + 1)] = '#';
    head = head->next;
  }

  for (int y = 1; y <= max_height; y++) {
    out[y * (max_width + 1) - 1] = '\n';
  }

  printf("%s\n", out);
}
